### PR TITLE
MCR-3164 add custom permission resolver to rest auth filter

### PR DESCRIPTION
**Pull request by golsch**
[Link to jira](https://mycore.atlassian.net/browse/MCR-3164).

### CODE REVIEW

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java
Comment (toKrause) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java]: Should be `new InternalServerErrorException(e);` so the original exception won't get lost where it is eventually logged/evaluated.

**Review by toKrause — CHANGES_REQUESTED** [review]

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java
Comment (Possommi) [reply mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java]: 159 `LogManager.getLogger().error(e);` should be removed since the exception is thrown

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/annotation/MCRRestRequiredPermission.java
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/annotation/MCRRestRequiredPermission.java:52]: This annotation is on a given endpoint, how does it have to resolve a permission (a.k.a. privilege) from the request? Can you elaborate a bit more on the use case?
For example if I write on a PUT endpoint @MCRRequiredPermission("myWritePerm") that's one permission I would want there. Where comes additional usage from the dynamic part in here?

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/access/MCRRestRequiredPermissionResolver.java
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/access/MCRRestRequiredPermissionResolver.java:34]: should add `throws WebApplicationException` as user provided value should be validated

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/access/MCRRestRequiredPermissionDefaultResolver.java
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/access/MCRRestRequiredPermissionDefaultResolver.java:27]: should always throw a InternalServerErrorException and the class should therefor be named MCRRestFailingPermissionResolver or similar

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java:181]: I think this could be simplified and have those cases reduced. See my other remarks

**Review by yagee-de — CHANGES_REQUESTED** [review]
I currently do not see the ContainerRequestContext to permission String use case yet.

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/annotation/MCRRestRequiredPermission.java
Comment (golsch) [reply mycore-restapi/src/main/java/org/mycore/restapi/v2/annotation/MCRRestRequiredPermission.java:52]: One case would be if the permission has to be derived from the properties of the resource itself. 
Assume there are read and write keys that are to be managed via one endpoint. For updating, different permissions are required.  If I want to update an access key, I do not know directly whether it is a read or write key and must first determine the corresponding permission.

Comment (yagee-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/v2/annotation/MCRRestRequiredPermission.java:52]: I see. Than I would suggest a quite different approach. A resolver should decide if the request is allowed or not. It should get the ResourceInfo and the ContainerRequestContext object and than make a final decision. A default implementation could than implement the current behavior of using the HTTP method and path parameters. This should come with a new Annotation then.

Comment (golsch) [reply mycore-restapi/src/main/java/org/mycore/restapi/v2/annotation/MCRRestRequiredPermission.java:52]: I had already thought about this.  However, it should always be ensured that the basic check resp. `#checkRestAPIAccess` is executed, which is why I opted for this approach. I had also considered refactoring `#checkRestAPIAccess` and then providing an abstract resolver, but I found the current approach  easier to use. Going further, I have considered a similar approach in combination with https://github.com/MyCoRe-Org/mycore/pull/2257.

**Comment by golsch** [comment]
Permissions should be set statically via the annotation and, if necessary, checked in more detail in a strategy. 
Example for read/write access keys, `MCRRestRequiredPermission("mange-access-keys")`. The final authorization with regard to read or write keys should then be checked in a strategy, see #2257.

**Pull request closed without merging**
